README.md

```diff
@@ -1 +1,419 @@
+
+```purescript
+route = root $ sum
+  { "Root": noArgs
+  , "Profile": path "user" (string segment)
+  , "Post": 
+      product 
+        (path "user" (string segment))
+        (path "post" (int segment))
+  , "Feed": path "feed" noArgs
+  }
+```
+
+It can be a little awkward using `product` for complex routes, so there’s also an operator version, `(/)`, which is more convenient to use infix. We can also omit the `string` combinator because all segments are strings by default. With this in mind, let’s revise our `”Post”` case:
+
+```purescript
+  {
+  , "Post": path "user" segment / path "post" (int segment)
+  }
+```
+
+In fact, when we’re matching string constants, we can omit the call to `path` and just provide the string directly:
+
+```purescript
+  {
+  , "Post": "user" / segment / "post" / int segment
+  }
+```
+
+## Example: Working with optional or required query params
+Users need to be able to search their feeds. This information will come via query parameters, which will be optional. We haven’t dealt with optional segments or query params so far, but they’re easy to add.
+
+First, let’s adjust our route type so that it can accommodate query parameters. Query parameters have a key:value pairing, so it’s typical to represent them with a record type.
+
+```purescript
+data Route
+  = Root
+  | Profile String
+  | Post String Int
+  | Feed { search :: Maybe String }
+```
+
+We’ll have to update our codec so that `Feed` takes a record as an argument. We can do this manually with the `record` function and its `:=` operator, which lets you assign a key in the record to a particular codec. Record keys are type-level strings, so we’ll need to use `SProxy` to create them.
+
+Intuitively, we can read the below codec as “Match `”feed”` and then, if it exists, a query parameter with the key “search”, storing its value at the key “search” in the output record.”
+
+```purescript
+route = root $ sum
+  { ...
+  , "Feed": path "feed" (record # _search := optional (param "search"))
+  }
+  where
+    _search = SProxy :: SProxy "search"
+```
+
+This explicit record creation can be done any time you have a record in your route type. However, using a record for query parameters is common enough that this library exports a helper function, `params`, which lets you just provide a record of codecs where the record keys are treated as the query param keys, too. We could rewrite our above codec using this helper function:
+
+```purescript
+  { ...
+  , "Feed": path "feed" $ params { search: optional <<< string }
+  -- alternately
+  , "Feed": "feed" / params { search: optional <<< string }
+  }
+```
+
+At this point, our codec is looking much cleaner:
+
+```purescript
+route = root $ sum
+  { "Root": noArgs
+  , "Profile": "user" / segment
+  , "Post": "user" / segment / "post" / int segment
+  , "Feed": "feed" / params { search: optional <<< string }
```

Comment:
Like `string` is unnecessary with `segment`, it's also unnecessary here. Maybe we could introduce `?` here as well now that it works with strings

```purescript
  , "Feed": "feed" ? { search: optional }
```

Reply:
👍